File: Data Structures in Python/HashMap.py
class HashNode:
    def __init__(self,key,value):
        self.key=key
        self.value = value
        self.next = None

class HashMap:
    def __init__(self,num):
        self.num=num # list的固定长度
        self.size =0 #一共有多少个值插入
        self.hash_table=self.createBucket()

    def createBucket(self):
        """
        生成固定长度的list，用来作为hash table的存放地址
        """
        hash_table = [None for i in range(self.num)]
        return hash_table

    def getBucketIndex(self,key):
        """
        :param key: int: 其key值
        :return index: int: 其key值在Hash Table这个array里的位置
        """
        hash_code=self.hashFunction(key)#首先得到从哈希函数中获得的键值
        index = hash_code % self.num #再通过取余数获得位置index
        return index

    def hashFunction(self,key):
        """
        :param key: int
        :return hash_code: int
        """
        hash_code = hash(key) #这里使用的是默认的哈希函数，也可以自己调整
        return hash_code

    def add(self,key,value):
        """
        :param key: int
        :param value: int
        :return: boolean
        核心方法：如何处理输入的数据，value是具体数据
        """
        bucket_index = self.getBucketIndex(key)
        head = self.hash_table[bucket_index] #获得头节点
        # 判断头节点是否为空，空则直接插入，非空则判断是否为key，是则修改数据
        while head:
            if head.key ==key:
                head.value = value
                return True
            else:
                head = head.next
        # 循环结束也没有找到匹配的值，或头结点为空，则插入新节点
        else:
            head = self.hash_table[bucket_index]
            new_node = HashNode(key,value)
            new_node.next = head
            self.hash_table[bucket_index] = new_node
            self.size+=1

        # 假如Array填满，需要更多的Array空间的话
        if self.size >=self.num:
            self.generateBiggerArray()

        return True

    def generateBiggerArray(self):
        """
        获得更大的array，将现有节点移动到bigger array里
        """
        temp = self.hash_table
        self.num = self.num * 2
        self.hash_table=self.createBucket()
        for each_head in temp:
            while each_head:
                self.add(each_head.key,each_head.value)
                each_head = each_head.next

    def get(self,key):
        """
        :param key: int
        :return: value: int or None
        """
        hash_index = self.getBucketIndex(key)
        head = self.hash_table[hash_index]
        while head:
            if head.key == key:
                return head.value
            else:
                head = head.next

        return None

    def getSize(self):
        return self.size

    def remove(self,key):
        hash_index = self.getBucketIndex(key)
        head = self.hash_table[hash_index]
        prev = None
        while head:
            if head.key == key:
                break
            prev = head
            head = head.next

        if head is None:
            return None

        if prev is not None:
            # head 不是头节点
            prev.next = head.next
        else:
            # head是头结点
            self.hash_table[hash_index]=head.next

        self.size -=1

        return head.value


def addAll(hm,key_list, value_list):
    for i in range(len(key_list)):
        hm.add(key_list[i],value_list[i])

def printAll(hm,key_list):
    for each in key_list:
        print(each,hm.get(each))

if __name__ =="__main__":
    key_list = [num for num in range(100)]
    value_list=[num for num in range(99,-1,-1)]
    hm = HashMap(20)
    addAll(hm,key_list,value_list)
    printAll(hm,key_list)

File: Data Structures in JAVA/HashMap.java
package DataStructure;


import java.util.*;
import java.util.Scanner;
public class HashMap {
    static class HashNode{
        /*哈希表节点*/
        String key;
        Integer value;
        HashNode next;
        public HashNode(String key, Integer value){
            this.key = key;
            this.value = value;
            this.next = null;
        }
    }
    private List<HashNode> bucketArray;
    private int numBuckets;//bucketArray长度
    private int size;//多少个元素
    public HashMap(){
        bucketArray =new ArrayList<>();
        numBuckets = 10;//简单定义为10 的案例
        size =0;
        for(int i=0;i<numBuckets;i++){
            bucketArray.add(null);//initialization -> bucketArray;
        }
    }
    private int getBucketIndex(String key){
        /*这里是哈希函数，把key值转换为哈希值*/
        int hashCode = key.hashCode();
        //JAVA虚拟机本身会自动给string直接生成hash code，不需要我们自己写
        int index = hashCode % numBuckets;//这里通过除MOD取余法计算index值
        return index;
    }
    public void add(String key, Integer value){
        int bucketIndex = getBucketIndex(key);
        HashNode head= bucketArray.get(bucketIndex);//获得位置index的元素
        //head是链地址法里链的头结点
        while(head!=null){
            if(head.key.equals(key)){
                //如果key值相同，意味着修改
                head.value = value;
                return;
            }
            head = head.next;
        }//循环结束也没有找到key，说明不存在该key的数据，进行添加
        head = bucketArray.get(bucketIndex);
        HashNode newNode = new HashNode(key,value);
        newNode.next = head;
        bucketArray.set(bucketIndex,newNode);
        size++;
        if((1.0*size)/numBuckets>=0.7){
            //保证存储的数据不要太多，导致冲突过多
            generateBiggerArray();
        }
    }

    private void generateBiggerArray(){
        /*扩容，将现有的bucketArray移到新的bucketArray上*/
        List<HashNode> temp = bucketArray;
        bucketArray = new ArrayList<>();
        numBuckets = 2 * numBuckets;
        for(int i=0;i<numBuckets;i++) {
            bucketArray.add(null);
        }
        for(HashNode headNode:temp){
            while(headNode!=null){
                add(headNode.key,headNode.value);
                headNode = headNode.next;
            }
        }
    }
    public Integer get(String key){
        /*查询函数*/
        int bucketIndex = getBucketIndex(key);
        HashNode head = bucketArray.get(bucketIndex);
        while(head!=null){
            if(head.key.equals(key)){
                return head.value;
            }
            head=head.next;
        }
        return null;
    }

    public Integer remove(String key){
        /*链表中删除节点的函数*/
        int bucketIndex = getBucketIndex(key);
        HashNode head = bucketArray.get(bucketIndex);
        HashNode prev = null;
        while(head!=null){
            if(head.key.equals(key)){
                break;
            }
            prev = head;
            head = head.next;
        }
        if(head == null){
            //没找到
            return null;
        }
        if(prev == null){
            //删除头结点
            bucketArray.set(bucketIndex, head.next);
        }
        else{
            prev.next = head.next;
        }
        size--;
        return head.value;
    }

    public boolean isEmpty() {
        return size==0;
}
    public static void main(String[] args){
        /*测试代码*/
        Scanner obj = new Scanner(System.in);
        System.out.println("Please enter the hash table node with format: key value \\n");
        HashMap hashMap=new HashMap();
        List<String> keyList = new ArrayList<>();
        while(obj.hasNextLine()){
            String strIn=obj.nextLine();
            String[] strList = strIn.split(" ");
            String key = strList[0];
            Integer value = Integer.parseInt(strList[1]);
            hashMap.add(key,value);
            keyList.add(key);
        }
        for(String each: keyList){
            System.out.println("the pair is: "+each+": "+hashMap.get(each));
        }
    }
}
